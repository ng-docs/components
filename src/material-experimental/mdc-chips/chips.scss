@use '@material/chips/deprecated' as mdc-chips;
@use '@material/ripple' as mdc-ripple;
@use '../../material/core/style/layout-common';
@use '../../cdk/a11y';
@use '../mdc-helpers/mdc-helpers';

@include mdc-chips.without-ripple($query: mdc-helpers.$mat-base-styles-query);
@include mdc-chips.set-core-styles($query: mdc-helpers.$mat-base-styles-query);

.mat-mdc-chip {
  // MDC uses a pointer cursor
  cursor: default;

  &._mat-animation-noopable {
    animation: none;
    transition: none;

    .mdc-chip__checkmark-svg {
      transition: none;
    }
  }

  @include a11y.high-contrast(active, off) {
    outline: solid 1px;

    &:focus {
      // Use 2px here since the dotted outline is a little thinner.
      outline: dotted 2px;
    }
  }
}

// The ripple container should match the bounds of the entire chip.
.mat-mdc-chip-ripple {
  @include layout-common.fill;

  // Disable pointer events for the ripple container and state overlay because the container
  // will overlay the user content and we don't want to disable mouse events on the user content.
  // Pointer events can be safely disabled because the ripple trigger element is the host element.
  pointer-events: none;

  // Inherit the border radius from the parent so that state overlay and ripples don't exceed the
  // parent button boundaries.
  border-radius: inherit;
}

// The MDC chip styles related to hover and focus states are intertwined with the MDC ripple styles.
// We currently don't use the MDC ripple due to size concerns, therefore we need to add some
// additional styles to restore these states.
.mdc-chip__ripple {
  @include mdc-ripple.target-common($query: structure);

  &::after, &::before {
    @include layout-common.fill;
    content: '';
    pointer-events: none;
    opacity: 0;
    border-radius: inherit;

    ._mat-animation-noopable & {
      transition: none;
    }
  }
}

// Angular Material supports disabled chips, which MDC does not.
// Dim the disabled chips and stop MDC from changing the icon color on click.
.mat-mdc-chip-disabled.mat-mdc-chip {
  opacity: 0.4;

  .mat-mdc-chip-trailing-icon, .mat-mdc-chip-row-focusable-text-content {
    pointer-events: none;
  }

  // Do not show state interactions for disabled chips.
  .mdc-chip__ripple::after, .mdc-chip__ripple::before {
    display: none;
  }
}

// Angular Material supports vertically-stacked chips, which MDC does not.
.mat-mdc-chip-set-stacked {
  flex-direction: column;
  align-items: flex-start;

  .mat-mdc-chip {
    width: 100%;
  }
}

// Add styles for the matChipInputFor input element.
$mat-chip-input-width: 150px;

input.mat-mdc-chip-input {
  flex: 1 0 $mat-chip-input-width;
}

// The margin value is set in MDC.
$chip-margin: 4px;

// Don't let the chip margin increase the mat-form-field height.
.mat-mdc-chip-grid {
  margin: -$chip-margin;

  // Keep the mat-chip-grid height the same even when there are no chips.
  input.mat-mdc-chip-input {
    margin: $chip-margin;
  }
}

.mdc-chip__checkmark-path {
  ._mat-animation-noopable & {
    transition: none;
  }

  @include a11y.high-contrast(black-on-white, off) {
    // SVG colors won't be changed in high contrast mode and since the checkmark is white
    // by default, it'll blend in with the background in black-on-white mode. Override the color
    // to ensure that it's visible. We need !important, because the theme styles are very specific.
    stroke: #000 !important;
  }
}

// Needed for the focus indicator.
.mat-mdc-chip-row-focusable-text-content {
  position: relative;
}

.mat-mdc-chip-remove {
  // Reset the user agent styles in case a button is used.
  border: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  padding: 0;
  background: none;

  .mat-icon {
    width: inherit;
    height: inherit;
    font-size: inherit;
  }
}

// Single-selection chips show their selected state using a background color which won't be visible
// in high contrast mode. This isn't necessary in multi-selection since there's a checkmark.
.mat-mdc-chip-selected:not(.mat-mdc-chip-multiple) {
  @include a11y.high-contrast(active, off) {
    outline-width: 3px;
  }
}

.mat-mdc-chip-row-focusable-text-content,
.mat-mdc-chip-remove-icon {
  display: flex;
  align-items: center;
}

.mat-mdc-chip-content {
  display: inline-flex;
}

.mdc-chip--editing {
  background-color: transparent;
  display: flex;
  flex-direction: column;

  .mat-mdc-chip-content {
    pointer-events: none;
    height: 0;
    overflow: hidden;
  }
}

.mat-chip-edit-input {
  cursor: text;
  display: inline-block;
}

.mat-mdc-chip-edit-input-container {
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
}
