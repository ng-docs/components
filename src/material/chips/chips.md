Chips allow users to view information, make selections, filter content, and enter data.

纸片允许用户查看信息、进行选择、过滤内容和输入数据。

### Static Chips

### 静态纸片

Chips are always used inside a container. To create chips, start with a `<mat-chip-set>` element. Then, nest `<mat-chip>` elements inside the `<mat-chip-set>`.

纸片总是在容器内使用。要创建纸片，请从 `<mat-chip-set>` 元素开始。然后，将 `<mat-chip>` 元素嵌套在 `<mat-chip-set>` 中。

<!-- example(chips-overview) -->

By default, `<mat-chip>` renders a chip with Material Design styles applied. For a chip with no styles applied, use `<mat-basic-chip>`.

默认情况下， `<mat-chip>` 会渲染应用了 Material Design 样式的纸片。对于未应用样式的纸片，请使用 `<mat-basic-chip>` 。

*Hint: `<mat-basic-chip>` receives the `mat-mdc-basic-chip` CSS class in addition to the `mat-mdc-chip` class.*

*提示： `<mat-basic-chip>` 除了 `mat-mdc-chip` 类之外还接收 `mat-mdc-basic-chip` CSS 类。*

#### Disabled appearance

#### 禁用态外观

Although `<mat-chip>` is not interactive, you can set the `disabled` Input to give it disabled appearance.

虽然 `<mat-chip>` 不是交互式的，但你可以设置 `disabled` 输入以使其具有禁用态外观。

```html
<mat-chip disabled>Orange</mat-chip>
```

### Selection Chips

### 选定纸片

Use `<mat-chip-listbox>` and `<mat-chip-option>` for selecting one or many items from a list. Start with creating a `<mat-chip-listbox>` element. If the user may select more than one option, add the `multiple` attribute. Nest a `<mat-chip-option>` element inside the `<mat-chip-listbox>` for each available option.

使用 `<mat-chip-listbox>` 和 `<mat-chip-option>` 从列表中选择一个或多个条目。从创建一个 `<mat-chip-listbox>` 元素开始。如果用户可以选择多个选项，请添加 `multiple` 属性。在 `<mat-chip-listbox>` 中为每个可用选项嵌套一个 `<mat-chip-option>` 元素。

#### Disabled `<mat-chip-option>`

#### 禁用 `<mat-chip-option>`

Use the `disabled` Input to disable a `<mat-chip-option>`. This gives the `<mat-chip-option>` a disabled appearance and prevents the user from interacting with it.

使用输入属性 `disabled` 来禁用 `<mat-chip-option>` 。这为 `<mat-chip-option>` 提供了禁用的外观，并阻止用户与其交互。

```html
<mat-chip-option disabled>Orange</mat-chip-option>
```

#### Keyboard Interactions

#### 键盘交互

Users can move through the chips using the arrow keys and select/deselect them with space. Chips also gain focus when clicked, ensuring keyboard navigation starts at the currently focused chip.

用户可以使用箭头键在纸片中移动，并使用空格选定/取消选定它们。单击时条状图也会获得焦点，确保键盘导航从当前聚焦的条状图开始。

### Chips connected to an input field

### 连接到输入字段的纸片

Use `<mat-chip-grid>` and `<mat-chip-row>` for assisting users with text entry.

使用 `<mat-chip-grid>` 和 `<mat-chip-row>` 来帮助用户输入文本。

Chips are always used inside a container. To create chips connected to an input field, start by creating a `<mat-chip-grid>` as the container. Add an `<input/>` element, and register it to the `<mat-chip-grid>` by passing the `matChipInputFor` Input. Always use an `<input/>` element with `<mat-chip-grid>`. Nest a `<mat-chip-row>` element inside the `<mat-chip-grid>` for each piece of data entered by the user. An example of using chips for text input.

纸片总是在容器内使用。要创建连接到输入字段的纸片，首先创建一个 `<mat-chip-grid>` 作为容器。添加一个 `<input/>` 元素，并通过把 `matChipInputFor` 传给 Input 将其注册到 `<mat-chip-grid>` 。始终将 `<input/>` 元素与 `<mat-chip-grid>` 一起使用。为用户输入的每条数据在 `<mat-chip-grid>` 中嵌套一个 `<mat-chip-row>` 元素。使用纸片进行文本输入的示例。

<!-- example(chips-input) -->

#### Disabled `<mat-chip-row>`

#### 禁用 `<mat-chip-row>`

Use the `disabled` Input to disable a `<mat-chip-row>`. This  gives the `<mat-chip-row>` a disabled appearance and prevents the user from interacting with it.

使用 `disabled` 的输入来禁用 `<mat-chip-row>` 。这为 `<mat-chip-row>` 提供了禁用的外观，并阻止用户与其交互。

```html
<mat-chip-row disabled>Orange</mat-chip-row>
```

#### Keyboard Interactions

#### 键盘交互

Users can move through the chips using the arrow keys and select/deselect them with the space. Chips also gain focus when clicked, ensuring keyboard navigation starts at the appropriate chip.

用户可以使用方向键在纸片之间移动，也可以用空格键选择它们或取消选择。 在点击时，纸片还会获得焦点，以确保会从合适的纸片开始导航。

Users can press delete to remove a chip. Pressing delete triggers the `removed` Output on the chip, so be sure to implement `removed` if you require that functionality.

用户可以按删除键删除纸片。按删除键会触发纸片上 `removed` 的输出属性，因此如果你需要该功能，请务必实现 `removed` 。

#### Autocomplete

#### 自动完成

An example of chip input with autocomplete.

具有自动完成功能的纸片输入示例。

<!-- example(chips-autocomplete) -->

### Icons

### 图标

You can add icons to chips to identify entities (like individuals) and provide additional functionality.

你可以向纸片添加图标以标识出实体（如个人）并提供附加功能。

#### Adding up to two icons with content projection

#### 添加最多两个带有内容投影的图标

You can add two additional icons to an individual chip. A chip has two slots to display icons using content projection. All variants of chips support adding icons including `<mat-chip>`, `<mat-chip-option>`, and `<mat-chip-row>`.

你可以向单个纸片添加两个额外的图标。一块纸片有两个插槽，可以使用内容投影来显示图标。所有纸片变体都支持添加图标，包括 `<mat-chip>` 、 `<mat-chip-option>` 和 `<mat-chip-row>` 。

A chip has a front slot for adding an avatar image. To add an avatar, nest an element with `matChipAvatar` attribute inside of `<mat-chip>`.

纸片有一个用于添加头像图像的正面插槽。要添加头像，请在 `<mat-chip>` 中嵌套一个具有 `matChipAvatar` 属性的元素。

<!-- example(chips-avatar) -->

You can add an additional icon to the back slot by nesting an element with either the `matChipTrailingIcon` or `matChipRemove` attribute.

你可以通过使用 `matChipTrailingIcon` 或 `matChipRemove` 属性嵌套元素来向后槽添加额外的图标。

#### Remove Button

#### 删除按钮

Sometimes the end user would like the ability to remove a chip. You can provide that functionality using `matChipRemove`. `matChipRemove` renders to the back slot of a chip and triggers the `removed` Output when clicked.

有时最终用户希望能够移除纸片。你可以使用 `matChipRemove` 提供该功能。 `matChipRemove` 会渲染到纸片的后插槽，并在单击时触发 `removed` 的输出。

To create a remove button, nest a `<button>` element with `matChipRemove` attribute inside the `<mat-chip-option>`. Be sure to implement the `removed` Output.

要创建移除按钮，请在 `<mat-chip-option>` 中嵌套一个具有 `matChipRemove` 属性的 `<button>` 元素。请务必实现 `removed` 输出属性。

```html
 <mat-chip-option>
  Orange
  <button matChipRemove aria-label="Remove orange">
    <mat-icon>cancel</mat-icon>
  </button>
</mat-chip-option>
```

See the [accessibility](#accessibility) section for best practices on implementing the `removed` Output and creating accessible icons.

有关如何创建无障碍图标的信息，请参阅[无障碍性](#accessibility)部分。

### Orientation

### 方向

By default, chips are displayed horizontally. To stack chips vertically, apply the `mat-mdc-chip-set-stacked` class to `<mat-chip-set>`, `<mat-chip-listbox>` or `<mat-chip-grid>`. 

默认情况下，纸片是水平显示的。要垂直堆叠纸片，请将 `mat-mdc-chip-set-stacked` 类应用于 `<mat-chip-set>` 、 `<mat-chip-listbox>` 或 `<mat-chip-grid>` 。

<!-- example(chips-stacked) -->

### Specifying global configuration defaults

### 指定全局配置默认值

Use the `MAT_CHIPS_DEFAULT_OPTIONS` token to specify default options for the chips module.

使用 `MAT_CHIPS_DEFAULT_OPTIONS` 令牌指定纸片模块的默认选项。

```html
@NgModule({
  providers: [
    {
      provide: MAT_CHIPS_DEFAULT_OPTIONS,
      useValue: {
        separatorKeyCodes: [COMMA, SPACE]
      }
    }
  ]
})
```

### Theming

### 主题

By default, chips use the primary color. Specify the `color` property to change the color to `accent` or `warn`.

默认情况下，纸片使用原色。指定 `color` 属性以将颜色更改为 `accent` 或 `warn` 。

### Interaction Patterns

### 交互模式

The chips components support 3 user interaction patterns, each with its own container and chip elements:

纸片组件支持 3 种用户交互模式，每种模式都有自己的容器和纸片元素：

#### Listbox

`<mat-chip-listbox>` and `<mat-chip-option>` : These elements implement a listbox accessibility pattern. Use them to present set of user selectable options.

`<mat-chip-listbox>` 和 `<mat-chip-option>` ：这些元素实现了列表框无障碍模式。使用它们来渲染一组可供用户选定的选项。

```html
<mat-chip-listbox aria-label="select a shirt size">
  <mat-chip-option> Small </mat-chip-option>
  <mat-chip-option> Medium </mat-chip-option>
  <mat-chip-option> Large </mat-chip-option>
</mat-chip-listbox>
```

#### Text Entry

#### 文字输入

`<mat-chip-grid>` and `<mat-chip-row>` : These elements implement a grid accessibility pattern. Use them as part of a free form input that allows users to enter text to add chips.

`<mat-chip-grid>` 和 `<mat-chip-row>` ：这些元素实现了网格无障碍性模式。将它们用作自由表单输入的一部分，允许用户输入文本以添加纸片。

```html
<mat-form-field>
  <mat-chip-grid #myChipGrid [(ngModel)]="mySelection"
  aria-label="enter sandwich fillings">
    <mat-chip-row *ngFor="let filling of fillings"
                 (removed)="remove(filling)">
      {{filling.name}}
      <button matChipRemove>
        <mat-icon>cancel</mat-icon>
      </button>
    </mat-chip-row>
    <input [matChipInputFor]="myChipGrid"
           [matChipInputSeparatorKeyCodes]="separatorKeysCodes"
           (matChipInputTokenEnd)="add($event)" />
  </mat-chip-grid>
</mat-form-field>
```

#### Static Content

#### 静态内容

`<mat-chip-set>` and `<mat-chip>` as an unordered list : Present a list of items that are not interactive. This interaction pattern mimics using `ul` and `li` elements. Apply role="list" to the `<mat-list>`. Apply role="listitem" to each `<mat-list-item>`.

`<mat-chip-set>` 和 `<mat-chip>` 作为无序列表：表示一个非交互式条目列表。这种交互模式模仿使用 `ul` 和 `li` 元素。将 role="list" 应用于 `<mat-list>` 。将 role="listitem" 应用于每个 `<mat-list-item>` 。

```html
<mat-chip-set role="list">
  <mat-chip role="listitem"> Sugar </mat-chip>
  <mat-chip role="listitem"> Spice </mat-chip>
  <mat-chip role="listitem"> Everything Nice </mat-chip>
</mat-chip-set>
```

`<mat-chip-set>` and `<mat-chip>` : These elements do not implement any specific accessibility pattern. Add the appropriate accessibility depending on the context. Note that Angular Material does not intend `<mat-chip>`, `<mat-basic-chip>`, and `<mat-chip-set>` to be interactive.

`<mat-chip-set>` 和 `<mat-chip>` ：这些元素不实现任何特定的无障碍模式。根据上下文添加适当的辅助功能。请注意，Angular Material 并不打算让 `<mat-chip>` 、 `<mat-basic-chip>` 和 `<mat-chip-set>` 进行交互。

```html
<mat-chip-set>
  <mat-chip> John </mat-chip>
  <mat-chip> Paul </mat-chip>
  <mat-chip> James </mat-chip>
</mat-chip-set>
```

### Accessibility

### 无障碍性

The [Interaction Patterns](#interaction-patterns) section describes the three variants of chips available. Choose the chip variant that best matches your use case.

[交互模式](#interaction-patterns)部分描述了可用纸片的三种变体。选择最适合你的用例的纸片变体。

For both MatChipGrid and MatChipListbox, always apply an accessible label to the control via `aria-label` or `aria-labelledby`.

对于 MatChipGrid 和 MatChipListbox，始终通过 `aria-label` 或 `aria-labelledby` 将无障碍标签应用于控件。

Always apply MatChipRemove to a `<button>` element, never a `<mat-icon>` element.

始终将 MatChipRemove 应用于 `<button>` 元素，而不是 `<mat-icon>` 元素。

When using MatChipListbox, never nest other interactive controls inside of the `<mat-chip-option>` element. Nesting controls degrades the experience for assistive technology users.

使用 MatChipListbox 时，切勿在 `<mat-chip-option>` 元素内嵌套其他交互式控件。嵌套控件会降低辅助技术用户的体验。

By default, `MatChipListbox` displays a checkmark to identify selected items. While you can hide the checkmark indicator for single-selection via `hideSingleSelectionIndicator`, this makes the component less accessible by making it harder or impossible for users to visually identify selected items.

When using `MatChipRemove`, you should always communicate removals for assistive technology. One way to accomplish this is by sending a message with `LiveAnnouncer`. Otherwise, removing a chip may only be communicated visually.

When a chip is editable, provide instructions to assistive technology how to edit the chip using a keyboard. One way to accomplish this is adding an `aria-description` attribute with instructions to press enter to edit the chip.
