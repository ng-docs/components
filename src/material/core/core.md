The `@angular/material/core` package contains common components and services used by multiple other
components in the library. See the API tab for a listing of components and classes available.
