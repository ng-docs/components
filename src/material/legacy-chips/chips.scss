@use '@angular/cdk';

@use '../core/style/variables';
@use '../core/style/elevation';
@use '../core/style/layout-common';
@use '../core/style/private' as style-private;
@use '../core/focus-indicators/private' as focus-indicators-private;

$chip-min-height: 32px;
$chip-vertical-padding: 7px;
$chip-horizontal-padding: 12px;

$chip-remove-vertical-padding: 7px;
$chip-remove-before-margin: 8px;
$chip-remove-after-padding: 8px;

$chip-avatar-vertical-padding: 0;
$chip-avatar-before-padding: 0;
$chip-avatar-before-margin: 4px;
$chip-avatar-after-margin: 8px;

$chip-margin: 4px;

$chip-input-width: 150px;
$chip-input-margin: 4px;

$chip-avatar-size: 24px;
$chip-remove-size: 18px;

.mat-chip {
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;

  // Chips could be set on buttons so we need to reset the user agent styles.
  border: none;
  -webkit-appearance: none;
  -moz-appearance: none;

  // For the chip element, default inset/offset values are necessary to ensure that
  // the focus indicator is sufficiently contrastive and renders appropriately.
  &::before {
    $border-width:
      var(--mat-focus-indicator-border-width, #{focus-indicators-private.$default-border-width});
    $offset: calc(#{$border-width} + 2px);
    margin: calc(#{$offset} * -1);
  }
}

.mat-standard-chip {
  @include elevation.elevation-transition;
  @include style-private.private-animation-noop;
  display: inline-flex;
  padding: $chip-vertical-padding $chip-horizontal-padding;
  border-radius: 16px;
  align-items: center;
  cursor: default;
  min-height: $chip-min-height;

  // Centering the content using flexbox won't work on IE, if we have
  // a `min-height` without setting a `height`. This height won't do
  // anything since it's less than the minimum set above.
  height: 1px;

  .mat-chip-remove {
    // Reset the user agent styles in case a button is used.
    border: none;
    -webkit-appearance: none;
    -moz-appearance: none;
    padding: 0;
    background: none;

    &.mat-icon, .mat-icon {
      width: $chip-remove-size;
      height: $chip-remove-size;
      font-size: $chip-remove-size;
    }
  }

  // Overlay used to darken the chip on hover and focus.
  &::after {
    @include layout-common.fill;
    border-radius: inherit;
    opacity: 0;
    content: '';
    pointer-events: none;
    transition: opacity 200ms variables.$swift-ease-in-out-timing-function;
  }

  &:hover::after {
    opacity: 0.12;
  }

  &:focus {
    outline: none;

    &::after {
      opacity: 0.16;
    }
  }

  @include cdk.high-contrast(active, off) {
    outline: solid 1px;

    // Seleted state is shown using a background color which isn't visible in high contrast mode.
    &.mat-chip-selected {
      outline-width: 3px;
    }
  }

  &.mat-chip-disabled {
    &::after {
      opacity: 0;
    }

    .mat-chip-remove,
    .mat-chip-trailing-icon {
      cursor: default;
    }
  }

  &.mat-chip-with-trailing-icon.mat-chip-with-avatar,
  &.mat-chip-with-avatar {
    padding-top: $chip-avatar-vertical-padding;
    padding-bottom: $chip-avatar-vertical-padding;
  }

  &.mat-chip-with-trailing-icon.mat-chip-with-avatar {
    padding-right: $chip-remove-after-padding;
    padding-left: $chip-avatar-before-padding;

    [dir='rtl'] & {
      padding-left: $chip-remove-after-padding;
      padding-right: $chip-avatar-before-padding;
    }
  }

  &.mat-chip-with-trailing-icon {
    padding-top: $chip-remove-vertical-padding;
    padding-bottom: $chip-remove-vertical-padding;
    padding-right: $chip-remove-after-padding;
    padding-left: $chip-horizontal-padding;

    [dir='rtl'] & {
      padding-left: $chip-remove-after-padding;
      padding-right: $chip-horizontal-padding;
    }
  }

  &.mat-chip-with-avatar {
    padding-left: $chip-avatar-before-padding;
    padding-right: $chip-horizontal-padding;

    [dir='rtl'] & {
      padding-right: $chip-avatar-before-padding;
      padding-left: $chip-horizontal-padding;
    }
  }

  .mat-chip-avatar {
    width: $chip-avatar-size;
    height: $chip-avatar-size;
    margin-right: $chip-avatar-after-margin;
    margin-left: $chip-avatar-before-margin;

    [dir='rtl'] & {
      margin-left: $chip-avatar-after-margin;
      margin-right: $chip-avatar-before-margin;
    }
  }

  .mat-chip-remove,
  .mat-chip-trailing-icon {
    width: $chip-remove-size;
    height: $chip-remove-size;
    cursor: pointer;
  }

  .mat-chip-remove,
  .mat-chip-trailing-icon {
    margin-left: $chip-remove-before-margin;
    margin-right: 0;

    [dir='rtl'] & {
      margin-right: $chip-remove-before-margin;
      margin-left: 0;
    }
  }
}

// Styles for the chip's dynamically added ripple target.
.mat-chip-ripple {
  @include layout-common.fill;

  // Disable pointer events for the ripple container and focus overlay because the container
  // will overlay the user content and we don't want to disable mouse events on the user content.
  // Pointer events can be safely disabled because the ripple trigger element is the host element.
  pointer-events: none;

  // Inherit the border radius from the parent so that ripples don't exceed the parent button
  // boundaries.
  border-radius: inherit;

  // Ensures that the ripple effect doesn't overflow the ripple target.
  overflow: hidden;

  // Required for the ripple to clip in Safari.
  transform: translateZ(0);
}

.mat-chip-list-wrapper {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  align-items: center;
  margin: -$chip-margin;

  input.mat-input-element,
  .mat-standard-chip {
    margin: $chip-margin;
  }
}

.mat-chip-list-stacked .mat-chip-list-wrapper {
  flex-direction: column;
  align-items: flex-start;

  .mat-standard-chip {
    width: 100%;
  }
}

.mat-chip-avatar {
  border-radius: 50%;
  justify-content: center;
  align-items: center;
  display: flex;
  overflow: hidden;

  // Makes `<img>` tags behave like `background-size: cover`. Not supported
  // in IE, but we're using it as a progressive enhancement.
  object-fit: cover;
}

input.mat-chip-input {
  width: $chip-input-width;
  margin: $chip-input-margin;
  flex: 1 0 $chip-input-width;
}
