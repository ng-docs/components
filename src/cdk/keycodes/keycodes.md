### KeyCodes

### 键码

Commonly used keycode constants.

常用的键码常量。

#### Example

#### 例子

```ts
import {Directive} from '@angular/core';
import {UP_ARROW, DOWN_ARROW, LEFT_ARROW, RIGHT_ARROW} from '@angular/cdk/keycodes';

@Directive({
  selector: '[count-arrows]'
  host: {
    (keypress): 'handleKeyPress($event)'
  }
})
export class ArrowCounterDirective {
  arrowPressCount = 0;

  handleKeyPress(event: KeyboardEvent) {
    if ([UP_ARROW, DOWN_ARROW, LEFT_ARROW, RIGHT_ARROW].includes(event.keyCode)) {
      this.arrowPresscount++;
    }
  }
}
```
