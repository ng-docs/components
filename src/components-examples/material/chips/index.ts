import {DragDropModule} from '@angular/cdk/drag-drop';
import {CommonModule} from '@angular/common';
import {NgModule} from '@angular/core';
import {ReactiveFormsModule} from '@angular/forms';
import {MatAutocompleteModule} from '@angular/material/autocomplete';
import {MatChipsModule} from '@angular/material/chips';
import {MatFormFieldModule} from '@angular/material/form-field';
import {MatIconModule} from '@angular/material/icon';
import {ChipsAutocompleteExample} from './chips-autocomplete/chips-autocomplete-example';
import {ChipsDragDropExample} from './chips-drag-drop/chips-drag-drop-example';
import {ChipsInputExample} from './chips-input/chips-input-example';
import {ChipsOverviewExample} from './chips-overview/chips-overview-example';
import {ChipsStackedExample} from './chips-stacked/chips-stacked-example';
import {ChipsHarnessExample} from './chips-harness/chips-harness-example';
import {ChipsFormControlExample} from './chips-form-control/chips-form-control-example';
import {ChipsAvatarExample} from './chips-avatar/chips-avatar-example';
import {MatButtonModule} from '@angular/material/button';

export {
  ChipsAutocompleteExample,
  ChipsDragDropExample,
  ChipsInputExample,
  ChipsOverviewExample,
  ChipsStackedExample,
  ChipsHarnessExample,
  ChipsFormControlExample,
  ChipsAvatarExample,
};

const EXAMPLES = [
  ChipsAutocompleteExample,
  ChipsDragDropExample,
  ChipsInputExample,
  ChipsOverviewExample,
  ChipsStackedExample,
  ChipsHarnessExample,
  ChipsFormControlExample,
  ChipsAvatarExample,
];

@NgModule({
  imports: [
    CommonModule,
    DragDropModule,
    MatAutocompleteModule,
    MatButtonModule,
    MatChipsModule,
    MatIconModule,
    MatFormFieldModule,
    ReactiveFormsModule,
  ],
  declarations: EXAMPLES,
  exports: EXAMPLES,
})
export class ChipsExamplesModule {}
