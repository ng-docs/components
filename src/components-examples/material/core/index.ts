import {NgModule} from '@angular/core';
import {FormsModule} from '@angular/forms';
import {MatButtonModule} from '@angular/material/button';
import {MatCheckboxModule} from '@angular/material/checkbox';
import {MatRippleModule} from '@angular/material/core';
import {MatInputModule} from '@angular/material/input';
import {ElevationOverviewExample} from './elevation-overview/elevation-overview-example';
import {RippleOverviewExample} from './ripple-overview/ripple-overview-example';

export {ElevationOverviewExample, RippleOverviewExample};

const EXAMPLES = [ElevationOverviewExample, RippleOverviewExample];

@NgModule({
  imports: [MatButtonModule, MatCheckboxModule, MatInputModule, MatRippleModule, FormsModule],
  declarations: EXAMPLES,
  exports: EXAMPLES,
})
export class CoreExamplesModule {}
