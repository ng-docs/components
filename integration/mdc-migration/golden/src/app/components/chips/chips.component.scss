.mat-mdc-chip {
  background-color: lavender;
  /* TODO(mdc-migration): The following rule targets internal classes of chips that may no longer apply for the MDC version. */
  .mat-chip-avatar {
    outline: 1px solid black;
  }
}


.mat-mdc-chip-grid { padding: 8px; }
