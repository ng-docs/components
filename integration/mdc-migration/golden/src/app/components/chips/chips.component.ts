import {COMMA, ENTER} from '@angular/cdk/keycodes';
import {Component} from '@angular/core';
import {MatChipInputEvent} from '@angular/material/chips';

export interface Fruit {
  name: string;
}

@Component({
  selector: 'chips-example',
  templateUrl: './chips.component.html',
  styleUrls: ['./chips.component.scss'],
})
export class ChipsComponent {
  addOnBlur = true;
  readonly separatorKeysCodes = [ENTER, COMMA] as const;
  fruits: Fruit[] = [{name: 'Lemon'}, {name: 'Lime'}, {name: 'Apple'}];

  add(event: MatChipInputEvent): void {
    const value = (event.value || '').trim();

    // Add our fruit
    if (value) {
      this.fruits.push({name: value});
    }

    // Clear the input value
    event.chipInput!.clear();
  }

  remove(fruit: Fruit): void {
    const index = this.fruits.indexOf(fruit);

    if (index >= 0) {
      this.fruits.splice(index, 1);
    }
  }
}
