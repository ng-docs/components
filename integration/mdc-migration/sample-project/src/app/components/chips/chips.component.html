<h2>Chips example</h2>

<mat-chip-list aria-label="Dog selection" #dogChipList>
    <mat-chip>
        <mat-chip-avatar>
            <img src="https://material.angular.io/assets/img/examples/shiba1.jpg" alt="Photo of a Shiba Inu"/>
        </mat-chip-avatar>
        Dog one
    </mat-chip>
    <mat-chip color="primary">Dog two</mat-chip>
    <mat-chip color="accent">Dog three</mat-chip>
</mat-chip-list>


<mat-form-field class="example-chip-list" appearance="fill">
    <mat-label>Favorite Fruits</mat-label>
    <mat-chip-list #fruitChipList aria-label="Fruit selection">
        <mat-chip *ngFor="let fruit of fruits" (removed)="remove(fruit)">
            {{fruit.name}}
            <button matChipRemove>
                <mat-icon>cancel</mat-icon>
            </button>
        </mat-chip>
        <input placeholder="New fruit..."
            [matChipInputFor]="fruitChipList"
            [matChipInputSeparatorKeyCodes]="separatorKeysCodes"
            [matChipInputAddOnBlur]="addOnBlur"
            (matChipInputTokenEnd)="add($event)">
    </mat-chip-list>
</mat-form-field>
