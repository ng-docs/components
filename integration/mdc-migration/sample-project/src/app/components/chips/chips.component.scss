.mat-chip {
  background-color: lavender;

  .mat-chip-avatar {
    outline: 1px solid black;
  }
}


.mat-chip-grid { padding: 8px; }
