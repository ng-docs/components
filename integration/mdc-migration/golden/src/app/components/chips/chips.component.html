<h2>Chips example</h2>

<mat-chip-listbox aria-label="Dog selection" #dogChipList>
    <mat-chip-option>
        <mat-chip-avatar>
            <img src="https://material.angular.io/assets/img/examples/shiba1.jpg" alt="Photo of a Shiba Inu"/>
        </mat-chip-avatar>
        Dog one
    </mat-chip-option>
    <mat-chip-option color="primary">Dog two</mat-chip-option>
    <mat-chip-option color="accent">Dog three</mat-chip-option>
</mat-chip-listbox>


<mat-form-field class="example-chip-list" appearance="fill">
    <mat-label>Favorite Fruits</mat-label>
    <mat-chip-grid #fruitChipList aria-label="Fruit selection">
        <mat-chip-row *ngFor="let fruit of fruits" (removed)="remove(fruit)">
            {{fruit.name}}
            <button matChipRemove>
                <mat-icon>cancel</mat-icon>
            </button>
        </mat-chip-row>
        <input placeholder="New fruit..."
            [matChipInputFor]="fruitChipList"
            [matChipInputSeparatorKeyCodes]="separatorKeysCodes"
            [matChipInputAddOnBlur]="addOnBlur"
            (matChipInputTokenEnd)="add($event)">
    </mat-chip-grid>
</mat-form-field>
